hugr-model/src/v0/scope/symbol.rs

```diff
@@ -0,0 +1,208 @@
+/// are always valid.
+///
+/// # Examples
+///
+/// ```
+/// # pub use hugr_model::v0::{NodeId, RegionId};
+/// # pub use hugr_model::v0::scope::SymbolTable;
+/// let mut symbols = SymbolTable::new();
+/// symbols.enter(RegionId(0));
+/// symbols.insert("foo", NodeId(0)).unwrap();
+/// assert_eq!(symbols.resolve("foo").unwrap(), NodeId(0));
+/// symbols.enter(RegionId(1));
+/// assert_eq!(symbols.resolve("foo").unwrap(), NodeId(0));
+/// symbols.insert("foo", NodeId(1)).unwrap();
+/// assert_eq!(symbols.resolve("foo").unwrap(), NodeId(1));
+/// assert!(!symbols.is_visible(NodeId(0)));
+/// symbols.exit();
+/// assert_eq!(symbols.resolve("foo").unwrap(), NodeId(0));
+/// assert!(symbols.is_visible(NodeId(0)));
+/// assert!(!symbols.is_visible(NodeId(1)));
+/// ```
+#[derive(Debug, Clone)]
+pub struct SymbolTable<'a> {
+    symbols: FxIndexMap<&'a str, BindingIndex>,
+    bindings: FxIndexMap<NodeId, Binding>,
+    scopes: FxIndexMap<RegionId, Scope>,
+}
+
+impl<'a> SymbolTable<'a> {
+    /// Create a new symbol table.
+    pub fn new() -> Self {
+        Self {
+            symbols: FxIndexMap::default(),
+            bindings: FxIndexMap::default(),
+            scopes: FxIndexMap::default(),
+        }
+    }
+
+    /// Enter a new scope for the given region.
+    pub fn enter(&mut self, region: RegionId) {
+        self.scopes.insert(
+            region,
+            Scope {
+                binding_stack: self.bindings.len(),
+            },
+        );
+    }
+
+    /// Exit a previously entered scope.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there are no remaining open scopes.
+    pub fn exit(&mut self) {
+        let (_, scope) = self.scopes.pop().unwrap();
+
+        for _ in scope.binding_stack..self.bindings.len() {
+            let (_, binding) = self.bindings.pop().unwrap();
+
+            if let Some(shadows) = binding.shadows {
+                self.symbols[binding.symbol_index] = shadows;
+            } else {
+                let last = self.symbols.pop();
+                debug_assert_eq!(last.unwrap().1, self.bindings.len());
+            }
+        }
+    }
+
+    /// Insert a new symbol into the current scope.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the symbol is already defined in the current scope.
+    /// In the case of an error the table remains unchanged.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there is no current scope.
+    pub fn insert(&mut self, name: &'a str, node: NodeId) -> Result<(), DuplicateSymbolError> {
+        let scope_depth = self.scopes.len() as u16 - 1;
+        let (symbol_index, shadowed) = self.symbols.insert_full(name, self.bindings.len());
+
+        if let Some(shadowed) = shadowed {
+            let (shadowed_node, shadowed_binding) = self.bindings.get_index(shadowed).unwrap();
+            if shadowed_binding.scope_depth == scope_depth {
+                self.symbols.insert(name, shadowed);
```

Comment:
If you used `get` and `insert`, the happy path would have two hash lookups. This way the happy path has just one and the error path has two. We could use the `entry` API to get down to one lookup in both paths; I did this initially and the code was quite unwieldy. Decided ultimately to go with this since we're stopping at an error anyway. I've only reinserted the symbol in the error path since APIs that leave inconconsistent state on error cause me nightmares.